File: src/senthil/model/electrolyte_potential.jl
include("conductivity.jl")

# Interfacial potentials
# Refer equations 105 & 106

"""Interfacial potential at negative electrode"""
function ϕ₂ᵢₙ(c₂ᵢₖ::InterfacialConc, q₂ᵢₖ::InterfacialFlux, I)
    2*θ*log.(c₂ᵢₖ.c₂ᵢₙ./c₂ₘₜ(c₂ᵢₖ, q₂ᵢₖ)) .+ I*lₛ./(2*k₂ₛ(c₂ᵢₖ, q₂ᵢₖ))
end

"""Interfacial potential at positive electrode"""
function ϕ₂ᵢₚ(c₂ᵢₖ::InterfacialConc, q₂ᵢₖ::InterfacialFlux, I)
    2*θ*log.(c₂ᵢₖ.c₂ᵢₚ./c₂ₘₜ(c₂ᵢₖ, q₂ᵢₖ)) .- I*lₛ./(2*k₂ₛ(c₂ᵢₖ, q₂ᵢₖ))
end

# Electrloyte potentials in different domains of battery
# Refer equations 104, 113 & 116

"""Electrloyte potential in negative electrode"""
function ϕ₂ₙ(x::Float64, c₂ᵢₖ::InterfacialConc, q₂ᵢₖ::InterfacialFlux, I)
    ϕ₂ᵢₙ(c₂ᵢₖ, q₂ᵢₖ, I) .+ 2*θ*log.(c₂ₙ(x, c₂ᵢₖ, q₂ᵢₖ)/c₂ᵢₖ.c₂ᵢₙ) .+ (I*(lₙ-x))./k₂ₙ(c₂ᵢₖ, q₂ᵢₖ) .- (I*(lₙ-x)^2)./(2*k₂ₙ(c₂ᵢₖ, q₂ᵢₖ)*lₙ)
end

"""Electrloyte potential in separator"""
function ϕ₂ₛ(x::Float64, c₂ᵢₖ::InterfacialConc, q₂ᵢₖ::InterfacialFlux, I)
    2*θ*log.(c₂ₛ(x, c₂ᵢₖ, q₂ᵢₖ)/c₂ₘₜ(c₂ᵢₖ, q₂ᵢₖ)) .- I*(x-(lₙ+lₛ/2))./k₂ₛ(c₂ᵢₖ, q₂ᵢₖ)
end

"""Electrloyte potential in positive electrode"""
function ϕ₂ₚ(x::Float64, c₂ᵢₖ::InterfacialConc, q₂ᵢₖ::InterfacialFlux, I)
    ϕ₂ᵢₚ(c₂ᵢₖ, q₂ᵢₖ, I) .+ 2*θ*log.(c₂ₚ(x, c₂ᵢₖ, q₂ᵢₖ)/c₂ᵢₖ.c₂ᵢₚ) .- (I*(x-lₙ-lₛ))./k₂ₚ(c₂ᵢₖ, q₂ᵢₖ) .+ (I*(x-lₙ-lₛ)^2)./(2*k₂ₚ(c₂ᵢₖ, q₂ᵢₖ)*lₚ)
end

# Overall electrolyte potential @x

"""Overall electrolyte potential function"""
function ϕ₂(x::Float64,c₂ᵢₖ::InterfacialConc, q₂ᵢₖ::InterfacialFlux, I)
    if x <= lₙ
        ϕ₂ₙ(x, c₂ᵢₖ, q₂ᵢₖ, I)
    elseif x > lₙ && x <= lₙ + lₛ
        ϕ₂ₛ(x, c₂ᵢₖ, q₂ᵢₖ, I)
    else
        ϕ₂ₚ(x, c₂ᵢₖ, q₂ᵢₖ, I)
    end
end

File: scripts/new_ashwini/model/electrolyte_potential.m
% Include initial_params.m
run('initial_params.m')

% initial_params

% Interfacial potential at negative electrode
function result = phi2in(I, params)
    result = 2*theta*log(c2in(params)./c2m(params)) + I*ls/(2*k2s(ln, params));
end

function result = phi2in(I, c2in, c2m)
    result = 2*theta*log(c2in./c2m) + I*ls/(2*k2s(c2in));
end

% Interfacial potential at positive electrode
function result = phi2ip(I, params)
    result = 2*theta*log(c2ip(params)./c2m(params)) - I*ls/(2*k2s(L-lp, params));
end

function result = phi2ip(I, c2ip, c2m)
    result = 2*theta*log(c2ip./c2m) - I*ls/(2*k2s(c2ip));
end

% Electrolyte potentials in different domains of battery

% TODO: Fix below 3 fns

% Electrolyte potential in negative electrode
function result = phi2n(x, I, params, varargin)
    dn1 = varargin{1};

    result = phi2in(I, params) + 2*theta*log(c2(x, params)./c2in(params)) + I*(ln-x)/k2n(x, params) - I*exp(dn1)*((ln-x) - ln*(1-exp(-dn1+dn1*x/ln))/dn1)/(k2n(x, params)*(exp(dn1) - 1));
end

function result = phi2n(x, I, c2, c2in, phi2in, dn1)
    result = phi2in + 2*theta*log(c2/c2in) + I*(ln-x)/k2n(c2) - I*exp(dn1)*((ln-x) - ln*(1-exp(-dn1+dn1*x/ln))/dn1)/(k2n(c2)*(exp(dn1) - 1));
end
